### PR TITLE
Add a warning for too large database row sizes

### PR DESCRIPTION
**Pull request by ausi**
To improve the situation for #4159 I created a small tool that is able to measure and calculate row sizes.

Not sure if this should ever get merged, but I needed a way to write the (somewhat confusing) rules for row sizes down and be able to validate them.

`getMysqlColumnSizeBits()` and `getInnodbColumnSizeBits()` calculates the amout of bits the column contributes to the row size.

`measureMysqlColumnSizeBits()` and `measureInnodbColumnSizeBits()` return the exact same information, but evaluate the amout of bits needed by trail and error against the database. (slow, only used for testing)

Bildschirmfoto 2022-02-19 um 23 46 42

### CODE REVIEW

**Comment by ausi** [comment]
As an example: in d4d4858c57643cd1c915683f40202a20505e0138 I reduced the row size of `tl_content` from 34698 down to 2663 (MySQL) and from 2313 down to 1998 (InnoDB).

**Side note:** Since MySQL 8.0.13 and MariaDB 10.2.1, default expressions are supported including for `TEXT` and `BLOB` columns. Are these versions too high as a system requirement for Contao 5.0?

File: core-bundle/src/Resources/contao/dca/tl_content.php [hunk 1]
```diff
@@ -194,7 +194,7 @@
 			'options_callback'        => array('tl_content', 'getContentElements'),
 			'reference'               => &$GLOBALS['TL_LANG']['CTE'],
 			'eval'                    => array('helpwizard'=>true, 'chosen'=>true, 'submitOnChange'=>true, 'tl_class'=>'w50'),
-			'sql'                     => array('name'=>'type', 'type'=>'string', 'length'=>64, 'default'=>'text')
+			'sql'                     => array('name'=>'type', 'type'=>'string', 'length'=>64, 'default'=>'text', 'platformOptions'=>array('collation'=>'ascii_bin'))
```
Comment (fritzmg) [inline core-bundle/src/Resources/contao/dca/tl_content.php]:
```suggestion
			'sql'                     => array('name'=>'type', 'type'=>'string', 'length'=>64, 'default'=>'text', 'customSchemaOptions'=>array('collation'=>'ascii_bin'))
```
As discussed in Slack :) https://www.doctrine-project.org/projects/doctrine-dbal/en/latest/reference/schema-representation.html#vendor-specific-options

**Review by fritzmg — CHANGES_REQUESTED** [review]

**Comment by ausi** [comment]
Briefly discussed with @fritzmg we should probably:

1. Apply all `COLLATE ascii_bin` changes
2. Change `varchar`s that are lager than 255 to `text`
3. Don’t touch all other `varchar`s as the change of the default value from `''` to `null` could cause backwards compatibility issues.

**Comment by aschempp** [comment]
could this _tool_ be useful to solve https://github.com/isotope/core/issues/1802 as well? 🙃

**Comment by ausi** [comment]
> could this _tool_ be useful to solve [isotope/core#1802](https://github.com/isotope/core/issues/1802) as well? 🙃

Don’t think so. The 64 key limit seems to be pretty straightforward. You should probably not create such a huge amount of indexes in the first place I would guess. Maybe some of them could be combined and some removed?

**Comment by ausi** [comment]
> 1. Apply all `COLLATE ascii_bin` changes
> 2. Change `varchar`s that are lager than 255 to `text`
> 3. Don’t touch all other `varchar`s as the change of the default value from `''` to `null` could cause backwards compatibility issues.

Done in 8db72fa4293aca0a2cf13f129710cf604e0ff494

Before this PR:

- `tl_content`: MySQL 34698 (53%) InnoDB 2313 (28%)
- `tl_module`: MySQL 25333 (39%) InnoDB 2726 (34%)

Now:

- `tl_content`: MySQL 14679 (22%) InnoDB 1963 (24%)
- `tl_module`: MySQL 5290 (8%) InnoDB 2427 (30%)

**Comment by ausi** [comment]
The changes from `tl_content` and `tl_module` should be backported to Contao 4.9.
@leofeyer should I make an extra PR after this one was approved?

File: core-bundle/src/Resources/contao/dca/tl_content.php [hunk 2]
```diff
@@ -277,30 +277,30 @@
 			'inputType'               => 'imageSize',
 			'reference'               => &$GLOBALS['TL_LANG']['MSC'],
 			'eval'                    => array('rgxp'=>'natural', 'includeBlankOption'=>true, 'nospace'=>true, 'helpwizard'=>true, 'tl_class'=>'w50'),
-			'sql'                     => "varchar(255) NOT NULL default ''"
+			'sql'                     => "varchar(64) COLLATE ascii_bin NOT NULL default ''"
```
Comment (fritzmg) [inline core-bundle/src/Resources/contao/dca/tl_content.php]: I am not sure about this change. If you have an image size called `this_lorem_ipsum_foobar_image_size` then the resulting serialized array

```
a:3:{i:0;i:0;i:1;i:0;i:2;s:23:"_this_lorem_ipsum_foobar_image_size";}
```

would be 69 characters long. (nice)

Comment (ausi) [reply core-bundle/src/Resources/contao/dca/tl_content.php]:
```suggestion
			'sql'                     => "varchar(128) COLLATE ascii_bin NOT NULL default ''"
```

File: core-bundle/src/Resources/contao/dca/tl_module.php [hunk 3]
```diff
@@ -435,21 +435,21 @@
 			'inputType'               => 'imageSize',
 			'reference'               => &$GLOBALS['TL_LANG']['MSC'],
 			'eval'                    => array('rgxp'=>'natural', 'includeBlankOption'=>true, 'nospace'=>true, 'helpwizard'=>true, 'tl_class'=>'w50'),
-			'sql'                     => "varchar(255) NOT NULL default ''"
+			'sql'                     => "varchar(64) COLLATE ascii_bin NOT NULL default ''"
```
Comment (ausi) [inline core-bundle/src/Resources/contao/dca/tl_module.php]:
```suggestion
			'sql'                     => "varchar(128) COLLATE ascii_bin NOT NULL default ''"
```

**Review by fritzmg — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
> The changes from `tl_content` and `tl_module` should be backported to Contao 4.9.
> @leofeyer should I make an extra PR after this one was approved?

Yes, please.

**Comment by leofeyer** [comment]
The changes do not work for me without https://github.com/ausi/contao-contao/pull/3. Contao will adjust the columns, however, since the `DcaSchemaProvider` does not handle the collation, the migration is never done.

**Review by leofeyer — CHANGES_REQUESTED** [review]
The changes do not work for me without https://github.com/ausi/contao-contao/pull/3. Contao will adjust the columns, however, since our `DcaSchemaProvider` does not handle collations, the migration is repeated over and over again.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @ausi.

**Pull request merged**